shapely/coordinates.py

```diff
@@ -54,20 +69,252 @@ def transform(geometry, transformation, include_z=False):
+    if rebuild:
+        transform_wrapper = (
+            transform_rebuild if interleaved else transform_rebuild_planar
+        )
+        vectorize = not isinstance(geometry, Geometry)
+    else:
+        transform_wrapper = transform if interleaved else transform_planar
+        vectorize = False
+        if not np.isscalar(include_z):
+            if np.all(include_z == include_z[0]):
+                include_z = include_z[0]
+            else:
+                vectorize = True
+    if vectorize:
+        transform_wrapper = np.frompyfunc(transform_wrapper, 3, 1)
+
+    return transform_wrapper(
+        geometry,
+        transformation,
+        include_z,
+    )
+
+
+def transform_planar(geometry, transformation, include_z: bool = False):
+    """Returns a copy of a geometry array with a function applied to its coordinates.
+
+    Refer to `shapely.transform_extra` (``rebuild=False``, ``interleaved=False``) for full documentation.
+    """
+    try:
+        # First we try to apply func to x, y, z vectors.
+        return transform(
+            geometry,
+            lambda coords: np.array(transformation(*coords.T)).T,
+            include_z=include_z,
+        )
+    except Exception:
+        # A func that assumes x, y, z are single values will likely raise a
+        # TypeError or a ValueError in which case we'll try again.
+        return transform(
+            geometry,
+            lambda coords: np.array([transformation(*c) for c in coords]),
+            include_z=include_z,
+        )
+
+
+def transform_rebuild(
+    geometry,
+    transformation,
+    include_z: bool = False,
+):
+    """Returns a copy of a geometry array with a function applied to its coordinates.
+
+    Refer to `shapely.transform_extra` (``rebuild=True``, ``interleaved=True``) for full documentation.
+    """
+    if geometry.is_empty:
+        return geometry
+    if geometry.geom_type in ("Point", "LineString", "LinearRing", "Polygon"):
+        return transform_rebuild_single_part(
+            geometry, transformation, include_z=include_z
+        )
+    elif (
+        geometry.geom_type.startswith("Multi")
+        or geometry.geom_type == "GeometryCollection"
+    ):
+        return type(geometry)(
+            [
+                transform_rebuild(part, transformation, include_z=include_z)
+                for part in geometry.geoms
+            ]
+        )
+    else:
+        raise GeometryTypeError(f"Type {geometry.geom_type!r} not recognized")
+
+
+def transform_rebuild_planar(
+    geometry,
+    transformation,
+    include_z: bool = False,
+):
+    """Returns a copy of a geometry array with a function applied to its coordinates.
+
+    Refer to `shapely.transform_extra` (``rebuild=True``, ``interleaved=False``) for full documentation.
+    """
+    if geometry.is_empty:
+        return geometry
+    if geometry.geom_type in ("Point", "LineString", "LinearRing", "Polygon"):
```

Comment:
We have a `GeometryType` enum, which should be used here (in combination with `shapely.get_type_id`. If you care for performance ;)